### PR TITLE
Refactor to generalize services

### PR DESCRIPTION
**Pull request by darinkrauss**
- Replace OSLog with DiagnosticLog
- Move GlucoseThreshold, LoopSettings, and TempBasalRecommendation from Loop
- Move BatteryStatus, IOBStatus, and PumpStatus from rileylink_ios
- Add Analytics, Logging, and RemoteData protocols
- Add Service and ServiceUI protocols
- Implement common ServiceTableViewController and ServiceViewController for simple services
- Implement MockService to record events from Analytics, Logging, and RemoteData procotols

[This is not up-to-date with the `dev` branch. I will include that next.]

### CODE REVIEW

**Comment by darinkrauss** [comment]
[Design new Service and ServiceUI protocols](https://trello.com/c/4hlXSYN6/695-design-new-service-and-serviceui-protocols)

**Comment by darinkrauss** [comment]
[Design new RemoteData, Logging, and Analytics protocols](https://trello.com/c/gvq8JOwr/694-design-new-remotedata-logging-and-analytics-protocols)

**Comment by darinkrauss** [comment]
[Refactor existing services to support new Service and ServiceUI protocols](https://trello.com/c/rUjClkL2/611-refactor-existing-services-to-support-new-service-and-serviceui-protocols)

**Comment by darinkrauss** [comment]
[Refactor service settings UI to support Service and ServiceUI](https://trello.com/c/XoV3ZGlJ/703-refactor-service-settings-ui-to-support-service-and-serviceui)

**Comment by darinkrauss** [comment]
[Refactor existing services to support new RemoteData, Logging, and Analytics protocols](https://trello.com/c/0jXmC3cG/704-refactor-existing-services-to-support-new-remotedata-logging-and-analytics-protocols)

File: LoopKit/LoopSettings.swift [hunk 1]
```diff
@@ -0,0 +1,108 @@
+//
+//  LoopSettings.swift
+//  Loop
+//
+//  Copyright © 2017 LoopKit Authors. All rights reserved.
+//
+
+
+public struct LoopSettings: Equatable {
```
Comment (ps2) [inline LoopKit/LoopSettings.swift]: This schema is a Loop specific thing, and shouldn't be in LoopKit. The Service API could provide a general "settings" store facility, perhaps key/value based. If the latter, some batching mechanism should be considered.

File: LoopKit/InsulinKit/TempBasalRecommendation.swift [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    public init(unitsPerHour: Double, duration: TimeInterval) {
+        self.unitsPerHour = unitsPerHour
+        self.duration = duration
+    }
+}
+
+
+extension TempBasalRecommendation {
```
Comment (ps2) [inline LoopKit/InsulinKit/TempBasalRecommendation.swift]: The functions in this extension are probably better kept in DoseMath as they are part of the Loop dosing algorithm.

File: LoopKit/Analytics.swift [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+//
+
+import Foundation
+
+
+public protocol Analytics {
+
+    // MARK: - UIApplicationDelegate
+
+    func application(didFinishLaunchingWithOptions launchOptions: [AnyHashable: Any]?)
```
Comment (ps2) [inline LoopKit/Analytics.swift]: This set of event types is a Loop app level concern, and probably shouldn't be in LoopKit.  If Service protocol will expose an analytics API, it probably needs to be generic to handle the changing set of event types.

File: LoopKit/PumpStatus.swift [hunk 4]
```diff
@@ -0,0 +1,30 @@
+//
+//  PumpStatus.swift
+//  LoopKit
+//
+//  Created by Darin Krauss on 5/21/19.
+//  Copyright © 2019 LoopKit Authors. All rights reserved.
+//
+
+
+public struct PumpStatus {
```
Comment (ps2) [inline LoopKit/PumpStatus.swift]: This part of the upload varies by pump. I'm wondering if we want an opaque Codable object for this.  The use cases it needs to support are 1) the hud widgets in Loop Follow, and 2) existing 'pills' in NightScout.

Some of the information here, like pump IOB, is not needed (remote monitoring tools should be displaying Loop IOB instead).  And some info, like `bolusing` and `suspended` should be covered by the other stream of dose information.

File: LoopKit/Service.swift [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+//  LoopKit
+//
+//  Created by Darin Krauss on 5/17/19.
+//  Copyright © 2019 LoopKit Authors. All rights reserved.
+//
+
+
+public protocol ServiceDelegate: AnyObject {
+
+    /// Informs the delegate that the specified service was updated.
```
Comment (ps2) [inline LoopKit/Service.swift:12]: What aspect of the service updated?  i.e. what would the delegate do in response to this message?

File: LoopKit/Service.swift [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+
+}
+
+
+public protocol Service: DeviceManager {
+
+    var serviceDelegate: ServiceDelegate? { get set }
+
+    /// Is the service complete?
+    var isComplete: Bool { get }
```
Comment (ps2) [inline LoopKit/Service.swift]: Does "complete" mean properly configured, or?

**Review by ps2 — CHANGES_REQUESTED** [review]

File: LoopKit/LoopSettings.swift (hunk 1 above)
Comment (darinkrauss) [reply LoopKit/LoopSettings.swift]: This is fixed. It was pulled in as needed by the Analytics protocol. With the update to the Analytics protocol to be more generic, this struct is no longer required in LoopKit and has been moved back to Loop.

File: LoopKit/InsulinKit/TempBasalRecommendation.swift (hunk 2 above)
Comment (darinkrauss) [reply LoopKit/InsulinKit/TempBasalRecommendation.swift]: Fixed. Extension is moved into DoseMath.

File: LoopKit/Analytics.swift (hunk 3 above)
Comment (darinkrauss) [reply LoopKit/Analytics.swift]: Fixed. The Analytics protocol has been made more general and the Loop-level analytics functions are fully contained within Loop.

File: LoopKit/PumpStatus.swift (hunk 4 above)
Comment (darinkrauss) [reply LoopKit/PumpStatus.swift]: This struct (and two others) were erroneously moved into LoopKit from rileylink_ios. They are now moved back to rileylink_ios.

File: LoopKit/Service.swift (hunk 5 above)
Comment (darinkrauss) [reply LoopKit/Service.swift:12]: Updated protocol documentation with further detail.

File: LoopKit/Service.swift (hunk 6 above)
Comment (darinkrauss) [reply LoopKit/Service.swift]: Renamed to `hasValidConfiguration` to be more descriptive.

**Comment by darinkrauss** [comment]
Stale, closing in favor of new PR.

**Pull request closed without merging**